lib/handler/mruby/middleware.c

```diff
@@ -0,0 +1,752 @@
+    if (name.base == NULL)
+        return 0;
+
+    if ((token = h2o_lookup_token(name.base, name.len)) != NULL) {
+        if (token == H2O_TOKEN_CONTENT_LENGTH) {
+            /* skip. use CONTENT_LENGTH instead of HTTP_CONTENT_LENGTH */
+        } else {
+            value = h2o_strdup(&req->pool, value.base, value.len);
+            h2o_add_header(&req->pool, &req->headers, token, NULL, value.base, value.len);
+        }
+    } else {
+        value = h2o_strdup(&req->pool, value.base, value.len);
+        h2o_add_header_by_str(&req->pool, &req->headers, name.base, name.len, 0, NULL, value.base, value.len);
+    }
+
+    return 0;
+}
+
+static void on_subreq_error_callback(void *data, h2o_iovec_t prefix, h2o_iovec_t msg)
+{
+    struct st_mruby_subreq_t *subreq = (void *)data;
+    mrb_state *mrb = subreq->ctx->shared->mrb;
+
+    assert(!mrb_nil_p(subreq->error_stream));
+
+    h2o_iovec_t list[] = {prefix, msg};
+    h2o_iovec_t concat = h2o_concat_list(&subreq->super.pool, list, 2);
+    mrb_value msgstr = h2o_mruby_new_str(mrb, concat.base, concat.len);
+    mrb_funcall(mrb, subreq->error_stream, "write", 1, msgstr);
+    if (mrb->exc != NULL) {
+        fprintf(stderr, "%s\n", RSTRING_PTR(mrb_inspect(mrb, mrb_obj_value(mrb->exc))));
+        mrb->exc = NULL;
+    }
+}
+
+static struct st_mruby_subreq_t *create_subreq(h2o_mruby_context_t *ctx, mrb_value env)
+{
+    static const h2o_conn_callbacks_t callbacks = {
+        get_sockname,    /* stringify address */
+        get_peername,    /* ditto */
+        NULL,            /* push (no push in subrequest) */
+        get_socket,      /* get underlying socket */
+        NULL,            /* get debug state */
+        {{{NULL}}}};
+
+    mrb_state *mrb = ctx->shared->mrb;
+    int gc_arena = mrb_gc_arena_save(mrb);
+    mrb_gc_protect(mrb, env);
+
+    /* create subreq */
+    struct st_mruby_subreq_t *subreq = h2o_mem_alloc(sizeof(*subreq));
+    h2o_req_t *super = &subreq->super;
+    memset(&subreq->conn, 0, sizeof(subreq->conn));
+    subreq->ctx = ctx;
+    subreq->receiver = mrb_nil_value();
+    subreq->ref = mrb_nil_value();
+    subreq->chunks = mrb_ary_new(mrb);
+    mrb_gc_register(mrb, subreq->chunks);
+    subreq->shortcut = NULL;
+    subreq->final_received = 0;
+    subreq->chain_proceed = 0;
+
+    /* initialize super and conn */
+    subreq->conn.super.ctx = ctx->shared->ctx;
+    h2o_init_request(&subreq->super, &subreq->conn.super, NULL);
+    super->is_subrequest = 1;
+    h2o_ostream_t *ostream = h2o_add_ostream(super, sizeof(*ostream), &super->_ostr_top);
+    ostream->do_send = subreq_ostream_send;
+    if (ctx->handler->pathconf->host) {
+        subreq->conn.super.hosts = h2o_mem_alloc_pool(&subreq->super.pool, sizeof(subreq->conn.super.hosts[0]) * 2);
+        subreq->conn.super.hosts[0] = ctx->handler->pathconf->host;
+        subreq->conn.super.hosts[1] = NULL;
+    } else {
+        subreq->conn.super.hosts = ctx->handler->pathconf->global->hosts;
+    }
```

Comment:
Shouldn't the subrequest inherit `h2o_conn_t::hosts` from the parent request?

`hosts` represents the host-level configurations that a request sent over the connection can be bound to.

Reply:
addressed at https://github.com/h2o/h2o/pull/1217/commits/4d7e72d8bdf4344bf6f4dfd40f1b18b60e744b02 and https://github.com/h2o/h2o/pull/1217/commits/e9ca27a4fd2d74efc57f0ce9326c71e8fb457a81

NOTE: we changed the behavior of proxy handler to always use globalconf->hosts, so we can  also use it for subreq->conn->hosts because all following handlers use it too

Reply:
https://github.com/h2o/h2o/commit/4d7e72d8bdf4344bf6f4dfd40f1b18b60e744b02 
causes a bug that `req->res.content_length` is not cleared when that request is replayed,. https://github.com/h2o/h2o/pull/1217/commits/a01f474811ba1465f99e7d080127659b16dc54b7 fixes that.